File: reference/python/pcam_runtime/rng.py
"""Canonical `pcam.pcg32.v1` deterministic random stream."""

from __future__ import annotations

from dataclasses import dataclass, replace

from .numeric import U64_MAX, apply_u64

PCG32_MULTIPLIER = 6364136223846793005


@dataclass(frozen=True)
class PCG32Stream:
    state: int
    stream_selector: int
    draw_count: int = 0
    algorithm_id: str = "pcam.pcg32.v1"

    @classmethod
    def seeded(cls, seed: int, stream_selector: int) -> "PCG32Stream":
        seed = apply_u64(seed)
        stream_selector = apply_u64(stream_selector)
        stream = cls(state=0, stream_selector=stream_selector)
        stream, _ = stream.draw_u32(count_draw=False)
        stream = replace(stream, state=(stream.state + seed) & U64_MAX)
        stream, _ = stream.draw_u32(count_draw=False)
        return stream

    @property
    def increment(self) -> int:
        return ((self.stream_selector << 1) | 1) & U64_MAX

    def draw_u32(self, count_draw: bool = True) -> tuple["PCG32Stream", int]:
        old_state = apply_u64(self.state)
        new_state = (old_state * PCG32_MULTIPLIER + self.increment) & U64_MAX
        xor_shifted = (((old_state >> 18) ^ old_state) >> 27) & 0xFFFFFFFF
        rotation = (old_state >> 59) & 31
        value = ((xor_shifted >> rotation) | (xor_shifted << ((-rotation) & 31))) & 0xFFFFFFFF
        return replace(self, state=new_state, draw_count=self.draw_count + int(count_draw)), value

    def to_snapshot(self) -> dict[str, int | str]:
        return {
            "algorithm_id": self.algorithm_id,
            "draw_count": self.draw_count,
            "state": self.state,
            "stream_selector": self.stream_selector,
        }

    @classmethod
    def from_snapshot(cls, snapshot: dict[str, object]) -> "PCG32Stream":
        if snapshot.get("algorithm_id") != "pcam.pcg32.v1":
            raise ValueError("RNG algorithm mismatch")
        return cls(
            state=apply_u64(int(snapshot["state"])),
            stream_selector=apply_u64(int(snapshot["stream_selector"])),
            draw_count=apply_u64(int(snapshot["draw_count"])),
        )

File: independent/rust/src/rng.rs
use serde::{Deserialize, Serialize};

pub const PCG32_MULTIPLIER: u64 = 6_364_136_223_846_793_005;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RngError {
    DrawCountOverflow,
    ProfileMismatch,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Pcg32Stream {
    pub algorithm_id: String,
    pub draw_count: u64,
    pub state: u64,
    pub stream_selector: u64,
}

impl Pcg32Stream {
    pub fn seeded(seed: u64, stream_selector: u64) -> Self {
        let mut stream = Self {
            algorithm_id: "pcam.pcg32.v1".to_owned(),
            draw_count: 0,
            state: 0,
            stream_selector,
        };
        stream.advance(false).expect("seeding does not count draws");
        stream.state = stream.state.wrapping_add(seed);
        stream.advance(false).expect("seeding does not count draws");
        stream
    }

    pub fn increment(&self) -> u64 {
        self.stream_selector.wrapping_shl(1) | 1
    }

    pub fn draw_u32(&mut self) -> Result<u32, RngError> {
        self.advance(true)
    }

    pub fn from_snapshot(snapshot: Self) -> Result<Self, RngError> {
        if snapshot.algorithm_id != "pcam.pcg32.v1" {
            return Err(RngError::ProfileMismatch);
        }
        Ok(snapshot)
    }

    fn advance(&mut self, count_draw: bool) -> Result<u32, RngError> {
        let old_state = self.state;
        let next_draw_count = if count_draw {
            self.draw_count
                .checked_add(1)
                .ok_or(RngError::DrawCountOverflow)?
        } else {
            self.draw_count
        };
        self.state = old_state
            .wrapping_mul(PCG32_MULTIPLIER)
            .wrapping_add(self.increment());
        let xor_shifted = (((old_state >> 18) ^ old_state) >> 27) as u32;
        let rotation = (old_state >> 59) as u32;
        if count_draw {
            self.draw_count = next_draw_count;
        }
        Ok(xor_shifted.rotate_right(rotation))
    }
}
